Makefile.rules

```diff
@@ -11,6 +11,11 @@ CFLAGS += -I../depends/include
 CFLAGS += -I../
 CFLAGS += "-DLOW_PRECISION=$(LOW_PRECISION)"
 CFLAGS += "-DHIGH_PRECISION=$(HIGH_PRECISION)"
+#This fixes linker errors on Mac.
+#Without it uninitialised variables are not included in static libraries.
+ifeq ($(shell uname -s),Darwin)
+    CFLAGS += "-fno-common"
```

Comment:
I find this a bit dangerous. It seems to suggest we have redudant definitions of global variables. See here:

http://www.keil.com/support/man/docs/armclang_ref/armclang_ref_chr1422532346348.htm

Reply:
Which variable were giving you trouble?

Reply:
It is because common block variables (by default non-initialised globals, in particular the All structure) are not exported by default in Mac's static library format (which is not ELF), so when it comes to linking the static library into the executable you get an unresolved error. This is only on Mac - it does not apply on linux at all.

The error was:
mpicc -fopenmp -O3 -g -Wall -ffast-math -march=native .objs/main.o .objs/params.o ../libgadget/libgadget.a ../libgadget/libgadget-utils.a -lm -L/usr/local/lib -lgsl -lgslcblas -lm -L../depends/lib -lmpsort-mpi -lradixsort -lbigfile-mpi -lbigfile -lpfft_omp -lfftw3_mpi -lfftw3_omp -lfftw3 -o MP-Gadget
Undefined symbols for architecture x86_64:
  "_All", referenced from:
      _main in main.o
      _OutputListAction in params.o
      _read_parameter_file in params.o
      _write_cpu_log in libgadget.a(run.o)
      _begrun in libgadget.a(run.o)
      _run in libgadget.a(run.o)
      _density_kernel_init in libgadget.a(densitykernel.o)
      ...
     (maybe you meant: _MPI_Alltoallv_smart, _MPI_Alltoallv_sparse )

Variables in the common block were:
All
Fd{CPU,BlackHole, etc}
NTask
ThisTask
ie, everything in allvars.c

Anything initialised to zero is exported, so one fix is to just change allvars.c to initialise everything to zero, as is done in, eg, partmanager.c Another fix would be to tell mac to explicitly export the common variables, but I wasn't sure how to do that without labelling all the variables, which is certain to break at some point.

What this does is to place the uninitialised variables in a block which is exported. Kind of a hack, but the least-invasive way to support Mac I could think of.

Reply:
At least this is not due to redundant definitions.

either adding -c to ranlib or -fno-common will work.

https://public.kitware.com/Bug/view.php?id=10342

Reply:
I think it is cleaner to have these in the local Makefile for OSX, rather than having too many if conditions in the main Makefile. We shall simply ship an example file that compiles cleanly on OSX.

Reply:
Ok, I'll commit that and merge this. Thanks.